Slim/Formats/Playlists/CUE.pm

```diff
@@ -60,144 +122,578 @@ sub parse {
+					value		=> $value,
+					filename	=> $filename,
+					returned	=> $1
+				}));
+
+			} elsif ($inAlbum and $value =~ /^\"?(\S+)\"?/i) {
+
+				# Some cue sheets may not have quotes. Allow that, but
+				# the filenames can't have any spaces in them."
+				$filename = $embedded || $1;
+				$filename = Slim::Utils::Misc::fixPath($filename, $baseDir);
+
+				$filesSeen++;
+				
+				main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump({
+					message		=> 'filename with no quotes',
+					line		=> $line,
+					command		=> $command,
+					value		=> $value,
+					filename	=> $filename,
+					returned	=> $1
+				}));
+
+			} elsif ($inAlbum){
+
+				# Invalid filename, skipped.
+				main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump({
+					message		=> 'Invalid filename',
+					line		=> $line,
+					command		=> $command,
+					value		=> $value
+				}));
+
+			} elsif (defined $currtrack and defined $filename){
+
+				# Better remove this point.
+				# To me it doesn't do what its was meant to do.
+				#
+				# In any case it will not take effect due to 
+				# the patch below marked : 
+				#
+				# Bug 5735, skip cue sheets with multiple FILE entries.
+				# 
+				# Here original comments:
+				#
+				# Each track in a cue sheet can have a different
+				# filename. See Bug 2126 &
+				# http://www.hydrogenaudio.org/forums/index.php?act=ST&f=20&t=4586
 			
-		} elsif (defined $currtrack and
+				$tracks->{$currtrack}->{'FILENAME'} = $filename;
+			} 
+			
+			# TODO: Correctly Handle Multiple file cue sheet.
+			# http://www.hydrogenaudio.org/forums/index.php?act=ST&f=20&t=4586
 
-			$line =~ /^(?:\s+REM )?\s*(TITLE|YEAR|GENRE|COMMENT|COMPOSER|CONDUCTOR|BAND)\s+\"(.*)\"/i) {
+		} else {
 
-			$tracks->{$currtrack}->{uc $1} = $2;
+			# handle remaning Commands as a list of keys and values.
+			($cuesheet, $tracks)=_addCommand($cuesheet, 
+											 $tracks,
+											 $inAlbum,
+											 $currtrack,
+											 $command,
+											 _removeQuotes($value));
+		}
+	}
+	main::DEBUGLOG && $log->is_debug && $log->debug(Data::Dump::dump({
+		message		=> 'after line parsing',
+		cuesheet	=> $cuesheet,
+		tracks		=> $tracks,
+		filename	=> $filename
+	}));
 
-		} elsif (defined $currtrack and
+	# Bug 5735, skip cue sheets with multiple FILE entries
+	if ( $filesSeen > 1 ) {
+		$log->warn('Skipping cuesheet with multiple FILE entries');
+		return;
+	}
 
-			$line =~ /^(?:\s+REM )?\s*(?:DISC|DISCNUMBER)\s+\"?([^\"]+)\"?/i) {
+	# Here controls on the entyre cuesheet structure, moving attributes
+	# to the correct level, preventing duplicates and renaming when needed.
+	#
+	if (defined $cuesheet->{'TITLE'}){
+		$cuesheet->{'ALBUM'} = $cuesheet->{'TITLE'};
+		delete $cuesheet->{'TITLE'};
+	}
 
-			# Some Cue sheets may not bother with quotes around a number
-			$tracks->{$currtrack}->{'DISC'} = $1;
+	if (defined $cuesheet->{'PERFORMER'}){
 
-		} elsif (defined $currtrack and
+		$cuesheet->{'ARTIST'} = $cuesheet->{'PERFORMER'};
```

Comment:
Standard Rule: the command Performer issued before any track command is the credited (album) artist, after a Track is the credited (track) artist.

Obviously this is too poor, so people is using workarounds.


Previous version:

REM ARTIST and REM ALBUM ARTIST at album level were discarded, so the only way to define ALBUM ARTIST (and ARTIST) was using PERFORMER.

Then if PERFORMER where used by tracks, ARTIST was populated (by track) and if not all tracks had the same Artist, then the album was pointed as COMPILATION.

Consider that after the parsing the Audio file is readed and if contains TAG (i.e Artist) they are 'imported' if not already defined, so making Artist = Album Artist = Performer is a sort of 'self protection' against errors coming from tags in audio file.

Other than this, LMS only consider track values, so at the end of parsing and after the tag reading from audio file, we have the merge into tracks of all the info detected by album, so Artist (equal album artist equal Performer) is the 'default' artist for tracks where performer (artist) is not defined in cue file.

Then, LMS looks for ALBUM ARTIST, ARTIST and TRACK ARTIST for each track and move AUDIO ARTIST (if unique) to the ALBUM level.

Why 3 different values? who knows...

This Mods:

Just accept any source, but keep the standard commands if defined over the others. You are sure you have just an Album Artist that, if defined, is the same as the Artist at album level.

Same fro tracks, If Artist is defined is the same of Track Artist, if Artist is different of Album artist, then the album is marked as compilation.

DIfferences:

1. Non considering values in non Audio Tracks (can cause false compilations to arise).
2. Test for 'true' value in Compilation, not just compilation is defined (but with a NO or 0 
value).
3. You could miss PERFORMER in cue sheet but have REM ALBUM ARTIST and/or REM ARTIST by tracks

What I'll like to do if not worried about use people already do of this metainfo:

LMS recognize 6 different Rules:

ALBUM ARTIST
TRACK ARTIST
ARTIST
COMPOSER
CONDUCTOR
BAND

ALBUM ARTIST, TRACK ARTIST togheter with COMPILATION has a specific meaning  and should be valued according with this use. ARTIST is somehow unclear, assuming dfferent meaning depending on where defined and if compilation or not, I'll avoid it at all in that sense, giving it the meaning of:

"Primary Performer" or Soloist, 

that togheter with 

COMPOSER
CONDUCTOR
BAND/ORCHESTRA

allows you to clearly tag your library for any genre. Now, since I could not use Artist for this, every performer, soloist, band, orchestra, chorus,... goes to BAND.

Now we MUST reserve PERFORMER to value ALBUM ARTIST and TRACK ARTIST but exactly as ARTIST it should be used to represent ...performers... that not in any situation is the same as Album artist or track Artist.

Why I did not modified in that sense:

a. people using LMS has in years found his way to use metainfo, any change will broke someone convenction.

b. COMPOSER is Songwriter in cue, but CONDUCTOR and BAND are unknown concept, you must use REM comments for them and other applications do not recognize them (Foobar just clear those lines from cue sheet when editing...).

c. same for ALBUM ARTIST and TRACK ARTIST, but in this case You should make some automate to value it from other metainfo if not found. From performer is an (arbitrary) way to do this.

What could be done:

If in some way is  possible to get a new preference and act according to that value I could produce a new version that could preserve the actual behaiour or handle separate meaning in a better way, is quite simple to do, but I don't Know if it is a viable solution.

Marco